keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+
+<!-- toc -->
+- [Summary](#summary)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Vocabulary](#vocabulary)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+      - [Admin](#admin)
+      - [User](#user)
+  - [APIs](#apis)
+    - [Storage APIs](#storage-apis)
+      - [BucketRequest](#bucketrequest)
+      - [Bucket](#bucket)
+      - [BucketClass](#bucketclass)
+    - [Access APIs](#access-apis)
+      - [BucketAccessRequest](#bucketaccessrequest)
+      - [BucketAccess](#bucketaccess)
+      - [BucketAccessClass](#bucketaccessclass)
+  - [Provisioner Secrets](#provisioner-secrets)
+<!-- /toc -->
+# Summary
+
+This proposal introduces the *Container Object Storage Interface* (COSI), a system composed of Custom Resource Definitions (CRDs), a controller architecture, and a gRPC specification, for the purpose of standardizing object storage representations in Kubernetes.  Goals and non-goals set the scope for the proposal by defining higher level objectives.  The vocabulary section defines terminology.  User stories illustrate how these APIs may fulfill cluster user requirements.  Relationships between the APIs are provided to illustrate the interconnections between object storage APIs, users' workloads, and object store service instances.  Lastly, the documents states the proposed API specs for the Bucket, Bucket, and BucketClass objects.
+
+## Motivation
+
+File and block are first class citizens within the Kubernetes ecosystem.  Object, though very different under the hood, is a popular means of storing data, especially against very large data sources.   As such, we feel it is in the interest of the community to integrate object storage into Kubernetes, supported by the SIG-Storage community.  In doing so, we can provide Kubernetes cluster users and administrators a normalized and familiar means of managing object storage. 
+
+While absolute portability cannot be guaranteed because of incompatibilities between providers, workloads reliant on a given protocol (e.g. one of S3, GCS, Azure Blob) may be defined in a single manifest and deployed wherever that protocol is supported.
+
+This proposal does _not_ include a standardized *protocol* or abstraction of storage vendor APIs
+
+## Goals
+
++ Specify object storage Kubernetes APIs for the purpose of orchestrating object store operations
++ Implement a Kubernetes controller architecture with support for pluggable provisioners
++ As MVP, be accessible to the largest groups of consumers by supporting the major object storage protocols (S3, Google Cloud Storage, Azure Blob) while being extensible for future protocol additions.
++ Present similar workflows for both greenfield and brownfield bucket operations.
+
+## Non-Goals
+
++ Define the _data-plane_ object store interface to replace or supplement existing vendor interfaces (i.e. replace GCS, S3, or Azure Blob)
+
+##  Vocabulary
+
++  _Brownfield Bucket_ - a storage instance created out of band.
```

Comment:
"an existing storage instance which could be created within or external to a Kubernetes cluster."